File: java-example/UsernameValidator.java
public class UsernameValidator {

    public boolean isValid(String username) {
        if (isTooShort(username)) {
            return false;
        }
        if (isTooLong(username)) {
            return false;
        }
        if (containsIllegalChars(username)) {
            return false;
        }
        return true;
    }

    boolean isTooShort(String username) {
        return username.length() < 3;
    }

    boolean isTooLong(String username) {
        return username.length() > 20;
    }

    // allows only alphanumeric and underscores
    boolean containsIllegalChars(String username) {
        return !username.matches("^[a-zA-Z0-9_]+$");
    }
}

File: python-example/username_validator.py
import re

class UsernameValidator:
    def is_valid(self, username):
        if self.is_too_short(username):
            return False
        if self.is_too_long(username):
            return False
        if self.contains_illegal_chars(username):
            return False
        return True
    
    def is_too_short(self, username):
        return len(username) < 3
    
    def is_too_long(self, username):
        return len(username) > 20
    
    # allows only alphanumeric and underscores
    def contains_illegal_chars(self, username):
        return not bool(re.match(r'^[a-zA-Z0-9_]+$', username))